usage/usage.go

```diff
@@ -0,0 +1,229 @@
+// unavailable.
+func Get() *api_schema.AgentResourceUsage {
+	agentResourceUsageMutex.Lock()
+	defer agentResourceUsageMutex.Unlock()
+
+	return agentResourceUsage
+}
+
+// Waits delay seconds, then starts polling resource usage every N seconds.  Use Get() to get the latest
+// usage data.
+func Poll(done <-chan struct{}, delay time.Duration, pollingInterval time.Duration) {
+	// Check if polling is disabled.
+	if pollingInterval <= 0 {
+		return
+	}
+
+	// Ensure only one thread is polling.
+	if !shouldStartPolling() {
+		return
+	}
+	setIsPolling(true)
+	defer setIsPolling(false)
+
+	history = queues.NewLinkedListQueue[statHistory]()
+
+	// Immediately record /proc state to compare against later.
+	stat, status, allStat, err := readProcFS()
+	if err != nil {
+		printer.Infof("Unable to poll for agent resource usage: %s\n", err)
+		return
+	}
+
+	history.Enqueue(statHistory{
+		stat:    stat,
+		status:  status,
+		allStat: allStat,
+	})
+
+	// Record usage after delay, then transition into recording every
+	// pollingInterval.
+	if delay >= 0 {
+		time.Sleep(delay)
+
+		if err := poll(pollingInterval); err != nil {
+			printer.Infof("Unable to poll for agent resource usage: %s\n", err)
+			return
+		}
+	}
+
+	// Record usage every pollingInterval.
+	ticker := time.NewTicker(pollingInterval)
+	for {
+		select {
+		case <-done:
+			return
+		case <-ticker.C:
+			if err := poll(pollingInterval); err != nil {
+				printer.Infof("Unable to poll for agent resource usage: %s\n", err)
+				return
+			}
+		}
+	}
+}
+
+func poll(pollingInterval time.Duration) error {
+	log := func(msg string, args ...any) {
+		if finishedFirstPoll {
+			return
+		}
+		printer.Infof(msg, args...)
+	}
+
+	observedAt := time.Now()
+	stat, status, allStat, err := readProcFS()
+	if err != nil {
+		return err
+	}
+
+	oldestStats, exist := history.Peek().Get()
+	if !exist {
+		// Should never happen.
+		return errors.Errorf("Internal error: missing resource usage history.")
+	}
+
+	lastStat := oldestStats.stat
+	lastAllStat := oldestStats.allStat
+
+	// Compute the processing time for this process vs. all processes since
+	// the last time Get() or Init() was called.
+	selfCPU := float64(stat.Utime-lastStat.Utime) + float64(stat.Stime-lastStat.Stime)
+	allCPU := float64(allStat.CPUStatAll.User-lastAllStat.CPUStatAll.User) +
+		float64(allStat.CPUStatAll.System-lastAllStat.CPUStatAll.System) +
+		float64(allStat.CPUStatAll.Idle-lastAllStat.CPUStatAll.Idle)
+	relativeCPU := selfCPU / allCPU
+	coresUsed := relativeCPU * float64(len(allStat.CPUStats))
+
+	// Get highest recorded VM high water mark.
+	vmHWM := status.VmHWM
+	history.ForEach(func(h statHistory) {
+		vmHWM = math.Max(vmHWM, h.status.VmHWM)
+	})
+
+	// Update history.
+	if history.Size() >= int(slidingWindowSize/pollingInterval) {
+		history.Dequeue()
+		history.Enqueue(statHistory{
+			stat:    stat,
+			status:  status,
+			allStat: allStat,
+		})
+	}
+
+	// Clear VM high-water mark.
```

Comment:
We want to high-water mark for the sliding window.  Resetting it after polling ensures we don't pick up a stale higher value.

Reply:
Isn't the global max RSS also a useful quantity to see, though? It's certainly the value we care about when we tune the CLI's memory usage. As it stands, we'll need to go through the entire history of reported measurements to get this value.

Reply:
That's true.  I'll add the global max values for CPU as well as memory to the telemetry.  I think it will also be useful to see recent usage over time along with the global max values.